pennylane/labs/tests/conftest.py

```diff
@@ -0,0 +1,192 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains shared fixtures for the device tests."""
+import argparse
+import os
+
+import numpy as np
+import pytest
+from _pytest.runner import pytest_runtest_makereport as orig_pytest_runtest_makereport
+
+import pennylane as qml
+
+# ==========================================================
+# pytest fixtures
+
+# seed for random functions
+np.random.seed(42)
+# Tolerance for analytic tests
+TOL = 1e-6
+# Tolerance for non-analytic tests
+TOL_STOCHASTIC = 0.05
+# Number of shots to call the devices with
+N_SHOTS = 1e6
+# List of all devices that are included in PennyLane
+LIST_CORE_DEVICES = {
+    "default.qubit",
+}
+
+
+@pytest.fixture(scope="function")
+def tol():
+    """Numerical tolerance for equality tests. Returns a different tolerance for tests
+    probing analytic or non-analytic devices, which allows us to define the
+    standard for deterministic or stochastic test results dynamically."""
+
+    def _tol(shots):
+        if shots is None:
+            return float(os.environ.get("TOL", TOL))
+        return TOL_STOCHASTIC
+
+    return _tol
+
+
+@pytest.fixture(scope="session")
+def init_state():
+    """Fixture to create an n-qubit random initial state vector."""
+
+    def _init_state(n):
+        state = np.random.random([2**n]) + np.random.random([2**n]) * 1j
+        state /= np.linalg.norm(state)
+        return state
+
+    return _init_state
+
+
+@pytest.fixture(scope="session")
+def skip_if():
+    """Fixture to skip tests."""
+
+    def _skip_if(dev, capabilities):
+        """Skip test if device has any of the given capabilities."""
+
+        dev_capabilities = dev.capabilities()

```

Comment:
Only legacy devices have a `capabilities()` method. New devices do not. As long as this fixture is only used with legacy devices, it should be fine, though we're slowly running out of legacy devices. Currently, pennylane only has 3 legacy devices:
* `default.mixed` (which is actually in the process of being ported to the new API)
* `default.qutrit`
* `default.gaussian`

Everything else, including `default.qubit`, uses the new API, and thus, doesn't have `capabilities` method.

Reply:
This may be out of scope for this PR to address, any ideas on how to move forward?

Reply:
Yeah, I agree, it seems out of scope. The way it's been written, it only affects legacy devices, so I think it's okay to leave it as is and assume that devs who use this fixture will only use it with legacy devices. We don't have a good way of validating device capabilities with the new device API yet.